### DISCUSSION TITLE
Creating scenicplus_obj

### DISCUSSION THREAD

**Discussion by Umaarasu** [aertslab/scenicplus#120] [General]
When I try to create the scenicplus obj ..I get the error Exception: No cells found which are present in both assays, check input and consider using `bc_transform_func`!. I checked the inputs and the barcodes match. Has someone had a similar issue?

I converted my seurat object for scRNA to Anndata for scATAC I followed the steps given here https://scenicplus.readthedocs.io/en/latest/faqs.html#i-have-an-analysis-with-another-tool-e-g-signac-archr-can-i-still-use-scenic

Code:

from scenicplus.scenicplus_class import create_SCENICPLUS_object
import numpy as np
scplus_obj = create_SCENICPLUS_object(
    GEX_anndata = adata,
    cisTopic_obj = cistopic_obj,
    menr = menr,
    bc_transform_func = lambda x: x  #function to convert scATAC-seq barcodes to scRNA-seq ones
)

Error: 
Exception                                 Traceback (most recent call last)
Cell In[108], line 3
      1 from scenicplus.scenicplus_class import create_SCENICPLUS_object
      2 import numpy as np
----> 3 scplus_obj = create_SCENICPLUS_object(
      4     GEX_anndata = adata,
      5     cisTopic_obj = cistopic_obj,
      6     menr = menr,
      7     bc_transform_func = lambda x: x  #function to convert scATAC-seq barcodes to scRNA-seq ones
      8 )
      9 scplus_obj.X_EXP = np.array(scplus_obj.X_EXP.todense())
     10 scplus_obj

File ~/scenicplus/src/scenicplus/scenicplus_class.py:533, in create_SCENICPLUS_object(GEX_anndata, cisTopic_obj, menr, multi_ome_mode, nr_metacells, nr_cells_per_metacells, meta_cell_split, key_to_group_by, imputed_acc_obj, imputed_acc_kwargs, normalize_imputed_acc, normalize_imputed_acc_kwargs, cell_metadata, region_metadata, gene_metadata, bc_transform_func, ACC_prefix, GEX_prefix)
    530 common_cells = list(set(GEX_cell_names) & set(ACC_cell_names))
    532 if len(common_cells) == 0:
--> 533     raise Exception(
    534         "No cells found which are present in both assays, check input and consider using `bc_transform_func`!")
    536 # impute accessbility if not given as parameter and subset for HQ cells
    537 if imputed_acc_obj is None:

Exception: No cells found which are present in both assays, check input and consider using `bc_transform_func`!

**Comment by SidG13** [comment]
Not sure if this will help you here, but check out [https://github.com/aertslab/scenicplus/issues/90](url). Generally you have a sample name which needs to be removed from one of the barcodes using the ```bc_transform_func```.

**Reply by Umaarasu** [reply]
Thanks for the suggestion.

My issue is the GEX data  is imported from seurat. Hence I do not have the  adata.raw, and for ATAC data I imported them from signac and processed them in cisTopic package as suggested in scenic FAQs. I am worried maybe this is difference is causing this issue. 

This is how my data looks and I can see the barcodes are similar and I don't get why the error shows up. 

AnnData object with n_obs × n_vars = 3448 × 36601
    obs: 'nCount_RNA', 'nFeature_RNA', 'percent.mt'
    var: 'name'

print(adata.obs['nCount_RNA'])
AAACATGCAAAGGTAC-1      76.0
AAACCAACACAACCTA-1      25.0

print(cistopic_obj.cell_data)
                                                 

                                                    cisTopic_log_nr_acc    Cell_type        cisTopic_log_nr_frag  \
AAACATGCAAAGGTAC-1            1.0000004                        EC                  4.1889   
AAACCAACACAACCTA-1            1.6000078                        ADl             4.003676

**Reply by SeppeDeWinter** [reply]
Hi

Could you show the output of

```python

len(set(cistopic_obj.cell_names) & set(adata.obs_names))

```

Best,

Seppe

**Reply by Umaarasu** [reply]
@SeppeDeWinter ..
I ran this 
" len(set(cistopic_obj.cell_names) & set(adata.obs_names))" and got 0.

When I checked 'cistopic_obj.cell_names'..it had the barcodes AAACATGCAAAGGTAC-1__cisTopic while my 'adata.obs_names' barcodes where  AAACATGCAAAGGTAC-1. I think this because I imported both my RNA and ATAC data from seurat, but processed the ATAC through cisTopic modelling (there was addition of __cisTopic to the barcodes).

I added __cisTopic to my barcodes at  'adata.obs_names'. That fixed the issue. 
"adata.obs_names = adata.obs_names + "___cisTopic""

P.S: It would be helpful if there is a tutorial showing how to import RNA data from a seurat object other than loom file. And what to expect when ATAC data is processed through cisTopic modelling and merged with RNA for SCENIC+.

Thanks!

**Reply by dcolinmorgan** [reply]
i used this function to split and remove `-10x-XXX` trailing text from the cistopic_obj cell names, then overwrite and re-export

```
def split_on_second_dash(strings):
    result = []
    for s in strings:
        parts = s.split('-')
        if len(parts) > 2:
            result.append('-'.join(parts[:2]))  # Join the first two parts
        else:
            result.append(s)  # If there are less than 2 dashes, return the original string
    return result
cistopic_obj.cell_names=split_on_second_dash(cistopic_obj.cell_names)
```

of course this just pushes the can down the road... one needs repeat this for `cistopic_obj.selected_model.cell_topic` and `cistopic_obj.cell_data.index`

**Comment by yojetsharma** [comment]
I was getting this error, so i changed the barcodes in my RNA h5ad file to match them in the cistopic_obj and around 51178 common barcodes intersected.
```
cistopic_barcodes:
 'AGTAACGAGCCTGGTA-1:WT___WT',
 'AGGAACCAGCGGATAA-1:WT___WT',
 'AGCGGATAGCAGGCCT-1:WT___WT',
 'TAGTAACGTACTGATG-1:WT___WT',
 'CAGGCTGGTCACCAAA-1:WT___WT',
 'GGGTTTGTCCATCAGG-1:WT___WT',
 'TGGTGATTCGCTAGCA-1:WT___WT',
 ...]
 snRNA_barcodes:
 Index(['AAACAGCCAAATATCC-1:WT___WT', 'AAACAGCCACAATGTT-1:WT___WT',
       'AAACAGCCAGTATGTT-1:WT___WT', 'AAACAGCCATAATCGT-1:WT___WT',
       'AAACAGCCATCCAGGT-1:WT___WT', 'AAACATGCAACTAGCC-1:WT___WT',
       'AAACATGCAACTGGGA-1:WT___WT', 'AAACATGCAAGCGATG-1:WT___WT',
       'AAACATGCACGAATCC-1:WT___WT', 'AAACATGCAGGCCATT-1:WT___WT',
```
How can I pass the bc_transform function such that no transformation takes place?
I tried "lambda x: x" or "x" but that didn't solve the issue either.

**Reply by SeppeDeWinter** [reply]
Hi @yojetsharma 

Indeed, `lambda x: x` should do the trick.

Can you share the full stack trace including command you are running so we can troubleshoot further?

All the best,

Seppe
